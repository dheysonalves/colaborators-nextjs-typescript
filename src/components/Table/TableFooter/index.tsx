import React, { useEffect } from 'react'
import {
	FooterContainer,
	FooterNumber,
	ButtonContainer,
	BackButtonContainer,
} from './styles'
import { DataProps } from '../../../hooks/useTable'
import { IoIosArrowForward, IoIosArrowBack } from 'react-icons/io'

interface TableFooterProps {
	range: number[]
	setPage: (value: number) => void
	page: number
	slice: DataProps[]
}

const TableFooter = ({ page, range, setPage, slice }: TableFooterProps) => {
	useEffect(() => {
		if (slice.length < 1 && page !== 1) {
			setPage(page - 1)
		}
	}, [slice, page, setPage])

	return (
		<FooterContainer>
			<BackButtonContainer
				onClick={() => setPage(page === 1 ? 1 : page - 1)}
				disabled={page === 1}
			>
				<IoIosArrowBack />
			</BackButtonContainer>
			<FooterNumber>
				{page} de {range.length}
			</FooterNumber>
			<ButtonContainer
				onClick={() => setPage(page === range.length ? page : page + 1)}
				disabled={page === range.length}
			>
				<IoIosArrowForward />
			</ButtonContainer>
		</FooterContainer>
	)
}

export default TableFooter
