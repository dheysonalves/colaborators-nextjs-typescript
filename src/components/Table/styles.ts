import styled from '@emotion/styled'

export const TableContainer = styled.table`
	border-collapse: collapse;
	border: none;
	width: 100%;
`

export const ThreadContainer = styled.thead`
	background-color: transparent;
	border: 1px solid #cad6d1;
`

export const ThreadHeader = styled.th`
	padding: 12px;
	font-weight: 600;
	font-size: 1.2rem;
	color: ${(props) => props.theme.light.neutral.neutral_5};
	text-align: left;

	:first-of-type {
		border-top-left-radius: 8px;
	}

	:last-of-type {
		border-top-right-radius: 8px;
	}
`

export const TableCell = styled.td`
	text-align: left;
	padding: 25px 12px;
	font-size: 1.2rem;
	font-weight: 400;
	color: ${(props) => props.theme.light.neutral.neutral_5};
	border-bottom: 1px solid ${(props) => props.theme.light.tertiary.disabled};
`
export const ImageCell = styled.img`
	width: 32px;
	height: 32px;
	object-fit: cover;
	border-radius: 50%;
`

export const NameImageWrapper = styled.div`
	display: flex;
	align-items: center;
`

export const NameSpan = styled.span`
	font-weight: 600;
	padding-left: 10px;
`
