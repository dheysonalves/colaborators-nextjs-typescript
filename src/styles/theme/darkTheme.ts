export const theme = {
	dark: {
		brand: {
			blue: '#034AFD',
		},
		neutral: {
			background: '#F8FAF9',
			light: '#fff',
			black: '#34423D',
			neutral_2: '#F5FAF8',
			neutral_3: '#A3B8B0',
			neutral_5: '#587169',
		},
		primary: {
			default: '#22E0A1',
			green: '#B5F1DD',
		},
		feedback: {
			success: '#B5F1DD',
		},
		tertiary: {
			disabled: '#EAEFED',
		},
	},
}
