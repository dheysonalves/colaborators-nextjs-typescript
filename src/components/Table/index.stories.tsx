import React from 'react'
import { ComponentMeta, ComponentStory } from '@storybook/react'
import Table from '.'

import {
	AgentStatusTypes,
	DepartamentRolesTypes,
	NameRolesTypes,
} from '../../api/@types/general.types'

export default {
	title: 'Composites/Table',
	component: Table,
	args: {
		data: [
			{
				id: 1,
				branch: 'Quartel General',
				department: DepartamentRolesTypes.ADMINISTRATIVE,
				name: 'Pikachu Soares do Santos Dias',
				role: NameRolesTypes.ANALYST,
				status: AgentStatusTypes.ACTIVE,
			},
			{
				id: 2,
				branch: 'Quartel General',
				department: DepartamentRolesTypes.ADMINISTRATIVE,
				name: 'Pikachu Soares do Santos Dias',
				role: NameRolesTypes.ANALYST,
				status: AgentStatusTypes.ACTIVE,
			},
			{
				id: 3,
				branch: 'Quartel General',
				department: DepartamentRolesTypes.ADMINISTRATIVE,
				name: 'Pikachu Soares do Santos Dias',
				role: NameRolesTypes.ANALYST,
				status: AgentStatusTypes.ACTIVE,
			},
			{
				id: 4,
				branch: 'Quartel General',
				department: DepartamentRolesTypes.ADMINISTRATIVE,
				name: 'Pikachu Soares do Santos Dias',
				role: NameRolesTypes.ANALYST,
				status: AgentStatusTypes.ACTIVE,
			},
			{
				id: 5,
				branch: 'Quartel General',
				department: DepartamentRolesTypes.ADMINISTRATIVE,
				name: 'Pikachu Soares do Santos Dias',
				role: NameRolesTypes.ANALYST,
				status: AgentStatusTypes.ACTIVE,
			},
		],
		rowsPerPage: 4,
	},
} as ComponentMeta<typeof Table>

export const Default: ComponentStory<typeof Table> = (args) => (
	<Table {...args} />
)
