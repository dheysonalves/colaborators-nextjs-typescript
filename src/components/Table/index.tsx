import React, { useState } from 'react'
import { BsThreeDotsVertical } from 'react-icons/bs'

import useTable from '../../hooks/useTable'
import TableFooter from './TableFooter'
import { DataProps } from '../../hooks/useTable'

import {
	TableContainer,
	ThreadContainer,
	ThreadHeader,
	TableCell,
	ImageCell,
	NameImageWrapper,
	NameSpan,
} from './styles'
import StatusPin from '../StatusPin/index'

interface TableProps {
	data: DataProps[]
}

const Table = ({ data }: TableProps) => {
	const [page, setPage] = useState<string | number>(1)
	const [rowsPerPage, setRowsPerPage] = useState<string | number>(data.length)
	const { range, slice } = useTable({ data, page, rowsPerPage })

	return (
		<>
			<TableContainer>
				<ThreadContainer>
					<tr>
						<ThreadHeader>Nome Completo</ThreadHeader>
						<ThreadHeader>Departamento</ThreadHeader>
						<ThreadHeader>Cargo</ThreadHeader>
						<ThreadHeader>Unidade</ThreadHeader>
						<ThreadHeader>Status</ThreadHeader>
					</tr>
				</ThreadContainer>
				<tbody>
					{slice.map((item) => (
						<tr key={item.agent_id}>
							<TableCell>
								<NameImageWrapper>
									<ImageCell src={item.image} />
									<NameSpan>{item.name}</NameSpan>
								</NameImageWrapper>
							</TableCell>
							<TableCell>{item.department}</TableCell>
							<TableCell>{item.role}</TableCell>
							<TableCell>{item.branch}</TableCell>
							<TableCell>
								<StatusPin status={item.status} />
							</TableCell>
							<TableCell>
								<BsThreeDotsVertical size={24} color="#587169" />
							</TableCell>
						</tr>
					))}
				</tbody>
			</TableContainer>
			<TableFooter
				range={range}
				slice={slice}
				setPage={setPage}
				page={page as number}
				rowsPerPage={rowsPerPage as number}
				setRowsPerPage={setRowsPerPage}
				dataLength={data.length}
			/>
		</>
	)
}

export default Table
