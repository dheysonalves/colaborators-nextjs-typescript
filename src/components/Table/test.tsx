import { ThemeProvider } from '@emotion/react'
import { render, screen } from '@testing-library/react'
import {
	AgentStatusTypes,
	DepartamentRolesTypes,
	NameRolesTypes,
} from '../../api/@types/general.types'
import { theme } from '../../styles/theme/theme'

import Table from '.'

const DATA = [
	{
		agent_id: 1,
		branch: 'Quartel General',
		department: DepartamentRolesTypes.ADMINISTRATIVE,
		name: 'Pikachu Soares do Santos Dias',
		role: NameRolesTypes.ANALYST,
		status: AgentStatusTypes.ACTIVE,
		image:
			'https://dailystatuss.com/wp-content/uploads/2022/02/273537712_1855045481347254_3794366435422454151_n.jpg',
	},
	{
		agent_id: 2,
		branch: 'Quartel General',
		department: DepartamentRolesTypes.ADMINISTRATIVE,
		name: 'Pikachu Soares do Santos Dias',
		role: NameRolesTypes.ANALYST,
		status: AgentStatusTypes.ACTIVE,
		image:
			'https://dailystatuss.com/wp-content/uploads/2022/02/273537712_1855045481347254_3794366435422454151_n.jpg',
	},
	{
		agent_id: 3,
		branch: 'Quartel General',
		department: DepartamentRolesTypes.ADMINISTRATIVE,
		name: 'Pikachu Soares do Santos Dias',
		role: NameRolesTypes.ANALYST,
		status: AgentStatusTypes.ACTIVE,
		image:
			'https://dailystatuss.com/wp-content/uploads/2022/02/273537712_1855045481347254_3794366435422454151_n.jpg',
	},
	{
		agent_id: 4,
		branch: 'Quartel General',
		department: DepartamentRolesTypes.ADMINISTRATIVE,
		name: 'Pikachu Soares do Santos Dias',
		role: NameRolesTypes.ANALYST,
		status: AgentStatusTypes.ACTIVE,
		image:
			'https://dailystatuss.com/wp-content/uploads/2022/02/273537712_1855045481347254_3794366435422454151_n.jpg',
	},
	{
		agent_id: 5,
		branch: 'Quartel General',
		department: DepartamentRolesTypes.ADMINISTRATIVE,
		name: 'Pikachu Soares do Santos Dias',
		role: NameRolesTypes.ANALYST,
		status: AgentStatusTypes.ACTIVE,
		image:
			'https://dailystatuss.com/wp-content/uploads/2022/02/273537712_1855045481347254_3794366435422454151_n.jpg',
	},
]

const setup = () => {
	const utils = render(
		<ThemeProvider theme={theme}>
			<Table data={DATA} />,
		</ThemeProvider>,
	)

	return {
		...utils,
	}
}

describe('<Table />', () => {
	it('should render the heading', () => {
		const { container } = setup()

		expect(screen.getByText('Nome Completo')).toBeInTheDocument()

		expect(container.firstChild).toMatchSnapshot()
	})
})
