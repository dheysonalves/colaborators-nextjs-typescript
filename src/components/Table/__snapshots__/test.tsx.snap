// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render the heading 1`] = `
<table
  class="css-1lngtp8"
>
  <thead
    class="css-1fdmb0w"
  >
    <tr>
      <th
        class="css-eqyatw"
      >
        Nome Completo
      </th>
      <th
        class="css-eqyatw"
      >
        Departamento
      </th>
      <th
        class="css-eqyatw"
      >
        Cargo
      </th>
      <th
        class="css-eqyatw"
      >
        Unidade
      </th>
      <th
        class="css-eqyatw"
      >
        Status
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        class="css-kssi1i"
      >
        <div
          class="css-ibdtyj"
        >
          <img
            class="css-ud1pw4"
            src="https://dailystatuss.com/wp-content/uploads/2022/02/273537712_1855045481347254_3794366435422454151_n.jpg"
          />
          <span
            class="css-scisbb"
          >
            Pikachu Soares do Santos Dias
          </span>
        </div>
      </td>
      <td
        class="css-kssi1i"
      >
        Administrativo
      </td>
      <td
        class="css-kssi1i"
      >
        Analista
      </td>
      <td
        class="css-kssi1i"
      >
        Quartel General
      </td>
      <td
        class="css-kssi1i"
      >
        <div
          class="css-1xl4ljt"
          color="active"
        >
          <h3
            class="css-1rt2vi7"
          >
            Ativo
          </h3>
        </div>
      </td>
      <td
        class="css-kssi1i"
      >
        <svg
          color="#587169"
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(88, 113, 105);"
          viewBox="0 0 16 16"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.5 13a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"
          />
        </svg>
      </td>
    </tr>
    <tr>
      <td
        class="css-kssi1i"
      >
        <div
          class="css-ibdtyj"
        >
          <img
            class="css-ud1pw4"
            src="https://dailystatuss.com/wp-content/uploads/2022/02/273537712_1855045481347254_3794366435422454151_n.jpg"
          />
          <span
            class="css-scisbb"
          >
            Pikachu Soares do Santos Dias
          </span>
        </div>
      </td>
      <td
        class="css-kssi1i"
      >
        Administrativo
      </td>
      <td
        class="css-kssi1i"
      >
        Analista
      </td>
      <td
        class="css-kssi1i"
      >
        Quartel General
      </td>
      <td
        class="css-kssi1i"
      >
        <div
          class="css-1xl4ljt"
          color="active"
        >
          <h3
            class="css-1rt2vi7"
          >
            Ativo
          </h3>
        </div>
      </td>
      <td
        class="css-kssi1i"
      >
        <svg
          color="#587169"
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(88, 113, 105);"
          viewBox="0 0 16 16"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.5 13a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"
          />
        </svg>
      </td>
    </tr>
  </tbody>
</table>
`
